mllib/src/main/scala/org/apache/spark/ml/reduction/OneVsRestClassifier.scala

```diff
@@ -0,0 +1,175 @@
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * Params for [[OneVsRestClassifier]].
+ */
+private[ml] trait OneVsRestParams extends ClassifierParams {
+
+  type ClassifierType = Classifier[F, E, M] forSome {
+    type F ;
+    type M <: ClassificationModel[F,M];
+    type E <:  Classifier[F, E,M]
+  }
+
+  /**
+   * param for the base classifier that we reduce multiclass classification into.
+   * @group param
+   */
+  val baseClassifier: Param[ClassifierType]  =
+    new Param(this, "baseClassifier", "base binary classifier/regressor ")
+
+  /** @group getParam */
+  def getBaseClassifier: ClassifierType = getOrDefault(baseClassifier)
+
+}
+
+/**
+ *
+ * @param parent
+ * @param baseClassificationModels the binary classification models for reduction.
+ */
+@AlphaComponent
+private[ml] class OneVsRestModel(
+    override val parent: OneVsRestClassifier,
+    val baseClassificationModels: Array[Model[_]])
+  extends Model[OneVsRestModel] with OneVsRestParams {
+
+  /**
+   * Transforms the dataset with provided parameter map as additional parameters.
+   * @param dataset input dataset
+   * @return transformed dataset
+   */
+  override def transform(dataset: DataFrame): DataFrame = {
+    // Check schema
+    val parentSchema = dataset.schema
+    transformSchema(parentSchema, logging = true)
+    val sqlCtx = dataset.sqlContext
+
+    // score each model on every data point and pick the model with highest score
+    val predictions = baseClassificationModels.zipWithIndex.par.map { case (model, index) =>
+      val output = model.transform(dataset)
+      output.select($(rawPredictionCol)).map { case Row(p: Vector) => List((index, p(1))) }
+    }.reduce[RDD[List[(Int, Double)]]] { case (x, y) =>
+      x.zip(y).map { case ((a, b)) =>
+        a ++ b
+      }
+    }.
+      map(_.maxBy(_._2))
+
+    // ensure that we pass through columns that are part of the original dataset.
+    val results = dataset.select(col("*")).rdd.zip(predictions).map { case ((row, (label, _))) =>
+      Row.fromSeq(row.toSeq ++ List(label.toDouble))
+    }
+
+    // the output schema should retain all input fields and add prediction column.
+    val outputSchema = SchemaUtils.appendColumn(parentSchema, $(predictionCol), DoubleType)
+    sqlCtx.createDataFrame(results, outputSchema)
+  }
+
+  @DeveloperApi
+  protected def featuresDataType: DataType = new VectorUDT
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema, fitting = false, featuresDataType)
+  }
+
+}
+
+/**
+ * :: Experimental ::
+ *
+ * Reduction of Multiclass Classification to Binary Classification.
+ * Performs reduction using one against all strategy.
+ * For a multiclass classification with k classes, train k models (one per class).
+ * Each example is scored against all k models and the model with highest score
+ * is picked to label the example.
+ *
+ */
+class OneVsRestClassifier extends Estimator[OneVsRestModel]
```

Comment:
Missing `@Experimental`. See my previous comments on the class name.

Should `OneVsRest` implement `Classifier`? @jkbradley